File: code/divide_conquer/src/inversion_count/inversion_count.cpp
//  divide conquer | inversion count | C++
//  Part of Cosmos by OpenGenus Foundation

#include <iostream>
using namespace std;

int mergesort(int arr[], int l, int r);
int merge(int arr[], int l, int m, int r);



int main()
{
    int n, a[100], i;
    cout << "Enter nuber of elements : ";
    cin >> n;
    cout << "Enter the elements : ";
    for (i = 0; i < n; i++)
        cin >> a[i];
    cout << "Number of inversion = " << mergesort(a, 0, n - 1);
    return 0;
}

//merge
int merge(int arr[], int l, int m, int r)
{
    int i, j, k, c[100], count = 0;
    i = 0;
    j = l;
    k = m;
    while (j <= m - 1 && k <= r)
    {
        if (arr[j] <= arr[k])
            c[i++] = arr[j++];

        else
        {
            c[i++] = arr[k++];
            count += m - j;

        }

    }

    while (j <= m - 1)
        c[i++] = arr[j++];

    while (k <= r)
        c[i++] = arr[k++];
    i = 0;
    while (l <= r)
        arr[l++] = c[i++];

    return count;
}


//mergesort
int mergesort(int arr[], int l, int r)
{
    int x = 0, y = 0, z = 0;
    int m = (l + r) / 2;
    if (l < r)
    {
        x += mergesort(arr, l, m);
        y += mergesort(arr, m + 1, r);
        z += merge(arr, l, m + 1, r);

    }

    return x + y + z;
}

File: code/divide_conquer/src/inversion_count/inversion_count.js
//  divide conquer | inversion count | Javascript
//  Part of Cosmos by OpenGenus Foundation

function mergeSort(arr, size) {
  if (array.length === 0 || array.length === 1) return 0;
  var temp = [];
  return mergeSortHelper(arr, temp, 0, size - 1);
}

function mergeSortHelper(arr, temp, left, right) {
  var inv_count,
    mid = 0;
  if (right > left) {
    mid = (right + left) / 2;

    /* sum the number of inversions from all parts */
    inv_count = mergeSortHelper(arr, temp, left, mid);
    inv_count += mergeSortHelper(arr, temp, mid + 1, right);

    inv_count += merge(arr, temp, left, mid + 1, right);
  }
  return inv_count;
}

function merge(arr, temp, left, mid, right) {
  var inv_count = 0;

  var i = left;
  var j = mid;
  var k = left;

  while (i <= mid - 1 && j <= right) {
    if (arr[i] <= arr[j]) {
      temp[k++] = arr[i++];
    } else {
      temp[k++] = arr[j++];

      /* remaining elements in subarray arr[i] to arr[mid] are sorted and grater than arr[j] */
      inv_count = inv_count + (mid - i);
    }
  }

  /* Copy remaining elements*/
  while (i <= mid - 1) temp[k++] = arr[i++];
  while (j <= right) temp[k++] = arr[j++];

  /*merge into original array*/
  for (i = left; i <= right; i++) arr[i] = temp[i];

  return inv_count;
}

console.log(mergeSort([5, 10, 4, 8, 2], 5));
